lib/TPP/CombineXsmmPass.cpp

```diff
@@ -0,0 +1,194 @@
+  // in BRGEMM.
+  int64_t strideA = lda * m;
+  int64_t strideB = ldb * k;
+  return DenseI64ArrayAttr::get(
+      rewriter.getContext(),
+      ArrayRef<int64_t>{m, n, k, lda, ldb, ldc, strideA, strideB});
+}
+
+ArrayAttr getUnaryFlags(RewriterBase &rewriter) {
+  return rewriter.getArrayAttr(
+      xsmm::UnaryFlagsAttr::get(rewriter.getContext(), xsmm::UnaryFlags::NONE));
+}
+
+ArrayAttr getBinaryFlags(RewriterBase &rewriter, xsmm::BinaryOp addOp) {
+  auto binaryInputType = addOp.getOperand(2).getType().cast<MemRefType>();
+  auto outputType = addOp.getOperand(3).getType().cast<MemRefType>();
+  return rewriter.getArrayAttr(xsmm::BinaryFlagsAttr::get(
+      rewriter.getContext(),
+      xsmm::utils::getBinaryBCast(binaryInputType, outputType,
+                                  /*operandNumber=*/0)));
+}
+
+xsmm::BinaryKindAttr getBinaryKind(RewriterBase &rewriter) {
+  return xsmm::BinaryKindAttr::get(rewriter.getContext(),
+                                   xsmm::BinaryKind::ADD);
+}
+
+xsmm::UnaryKindAttr getUnaryKind(RewriterBase &rewriter) {
+  return xsmm::UnaryKindAttr::get(rewriter.getContext(), xsmm::UnaryKind::RELU);
+}
+
+static ArrayAttr getGemmFlags(RewriterBase &rewriter, xsmm::BrgemmOp opTy) {
+  auto memrefB = opTy.getOperand(2).getType().cast<MemRefType>();
+  xsmm::GemmFlagsAttr gemmFlag =
+      (vnni::utils::isInVnniLayout(memrefB))
+          ? xsmm::GemmFlagsAttr::get(rewriter.getContext(),
+                                     xsmm::GemmFlags::VNNI_B)
+          : xsmm::GemmFlagsAttr::get(rewriter.getContext(),
+                                     xsmm::GemmFlags::NONE);
+  return rewriter.getArrayAttr(gemmFlag);
+}
+
+struct CombineXsmmOp : public OpRewritePattern<xsmm::BrgemmOp> {
+  using OpRewritePattern<xsmm::BrgemmOp>::OpRewritePattern;
+  LogicalResult matchAndRewrite(xsmm::BrgemmOp brgemmOp,
+                                PatternRewriter &rewriter) const override {
+    auto brgemmResultOperand = brgemmOp.getOperand(3);
+
+    for (auto use = brgemmResultOperand.user_begin();
+         use != brgemmResultOperand.user_end(); use++) {
+      auto user = *use;
+      if (isa<xsmm::BinaryOp>(user) &&
+          (dyn_cast<xsmm::BinaryOp>(user)).getCallee() ==
+              mlir::xsmm::BinaryKind::ADD) {
+        xsmm::BinaryOp binaryOp = ((xsmm::BinaryOp)user);
+        auto dispatchOp = dyn_cast<xsmm::BinaryDispatchOp>(
+            binaryOp.getOperand(0).getDefiningOp());
+        auto isFusedAdd = dispatchOp.getKind() == xsmm::BinaryKind::ADD;
+        auto binaryFlags = dispatchOp.getFlags();
+        if (isFusedAdd &&
+            (binaryFlags.size() != 1 ||
+             binaryFlags[0].cast<mlir::xsmm::BinaryFlagsAttr>().getValue() !=
+                 mlir::xsmm::BinaryFlags::BCAST_COL_IN_0)) {
+          return failure();
+        }
+        auto addResultOperand = ((xsmm::BinaryOp)user).getOperand(3);
+        for (auto relu = addResultOperand.user_begin();
```

Comment:
Do not have loops inside loops. This won't scale for anything that is not an ADD followed by a RELU.

As we discussed earlier, have an initial scan for binary/unary, put them in a list/struct/local vars, and then handle the generic case from there.

Reply:
will address in a subsequent commit

Reply:
No, this will create technical debt that will be hard to remove. We can't go on adding things that will be removed later.